Notifications API actions/maxActions and friends (#44184)
mdn/content: files/en-us/web/api/notification/actions/index.md

{{APIRef("Web Notifications")}}{{SecureContext_Header}}{{SeeCompatTable}} {{AvailableInWorkers}}

The **`actions`** read-only property of the {{domxref("Notification")}} interface provides the actions available for users to select when interacting with the notification.

## Value

A read-only array of actions.
Each element in the array is an object with the following members:

> - `action`

Reviewer: The other object members make sense, but this one feels like it needs more information. What actions are available to be specified? Are they specified somewhere, and/or can a developer specify their own custom actions? If this info is available somewhere else, a link would suffice.

Author: Good point. Fixed in https://github.com/mdn/content/pull/44184/changes/616e66278c381917b5a1d02f52930b9f7c9f3b8c

Reviewer: Perfect.

  - : A string identifying a user action to be displayed on the notification.
- `title`
  - : A string containing action text to be shown to the user.
- `icon`
  - : A string containing the URL of an icon to display with the action.

> - `navigate` {{optional_inline}} {{experimental_inline}}

Author: FYI in spec, and can be set - though not on FF

  - : A string containing a URL to navigate to when the user activates this action.
    When set, the user agent navigates to this URL instead of firing the {{domxref("ServiceWorkerGlobalScope.notificationclick_event", "notificationclick")}} event.
    See {{domxref("Notification.navigate")}} for more information.

## Description

Notification actions are buttons or controls that appear within [persistent notifications](/en-US/docs/Web/API/Notifications_API#persistent_and_non-persistent_notifications).
They are set using the [`actions`](/en-US/docs/Web/API/ServiceWorkerRegistration/showNotification#actions) option of the second argument of the {{domxref("ServiceWorkerRegistration.showNotification", "showNotification()")}} method.

> Note that actions are not available for non-persistent notifications: a `TypeError` is thrown if you pass an options object with a non-`null` actions property to the {{domxref("Notification/Notification", "Notification()")}} constructor.

Reviewer: ```suggestion
Note that actions are not available for non-persistent notifications: In non-persistent cases, a `TypeError` is thrown if you pass an options object with a non-`null` actions property to the {{domxref("Notification/Notification", "Notification()")}} constructor.
```

Author: Using `Notification()` is what makes the notification non-persistent - with this change it sounds like there might be persistent and non-persistent notifications created using `Notification()`. 

Perhaps inverting the form will make that more clear.

```suggestion
Note that actions are not available for non-persistent notifications.
If you pass an `options` object with an `actions` property that is anything other than `null` to the {{domxref("Notification/Notification", "Notification()")}} constructor, a `TypeError` is thrown.
```

Reviewer: Yup, better.

Clicking the button associated with an action navigates to the URL set in the [`navigate`](#navigate) option if one is specified.

> Otherwise it fires a [`notificationclick`](/en-US/docs/Web/API/ServiceWorkerGlobalScope/notificationclick_event) event on the service worker that includes the action that was selected (and the associated `Notification` instance), so the worker can handle it without the user ever switching to your page.

Reviewer: ```suggestion
Otherwise, it fires a [`notificationclick`](/en-US/docs/Web/API/ServiceWorkerGlobalScope/notificationclick_event) event on the service worker that includes the action that was selected (and the associated `Notification` instance), so the worker can handle it without the user ever switching to your page.
```

> [!NOTE]
> Browsers typically limit the maximum number of actions they will display for a particular notification.
> Check the static {{domxref("Notification.maxActions_static", "Notification.maxActions")}} property to determine the limit.

## Examples

### Basic usage

> The following code shows how a service worker might listen for the `notificationclick` event and use it to get both the clicked action and all the actions.

Reviewer: ```suggestion
The following code shows how a service worker might listen for the `notificationclick` event and use it to retrieve both the clicked action and an array of all actions.
```
